javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll

```diff
@@ -161,6 +161,60 @@ module DataFlow {
 
     /** Gets a textual representation of this element. */
     string toString() { none() }
+
+    /**
+     * Gets the immediate predecessor of this node, if any.
+     *
+     * A node with an immediate predecessor can only ever have the value that flows
+     * into its from its immediate predecessor. It can be treated as an alias for
+     * the immediate predecessor.
+     */
+    cached
+    DataFlow::Node getImmediatePredecessor() {
+      // Use of variable -> definition of variable
+      exists(SsaVariable var |
+        this = DataFlow::valueNode(var.getAUse()) and
+        result.(DataFlow::SsaDefinitionNode).getSsaVariable() = var
+      )
+      or
+      // Refinement of variable -> original definition of variable
+      exists(SsaRefinementNode refinement |
+        this.(DataFlow::SsaDefinitionNode).getSsaVariable() = refinement.getVariable() and
+        result.(DataFlow::SsaDefinitionNode).getSsaVariable() = refinement.getAnInput()
+      )
+      or
+      // Definition of variable -> RHS of definition
+      exists(SsaExplicitDefinition def |
+        this = TSsaDefNode(def) and
+        result = def.getRhsNode()
```

Comment:
In the presence of default values in destructuring assignments, a single definition can have multiple right-hand sides. We'd either have to `strictcount` that here, or introduce a new data-flow node that merges the right-hand sides.